pymc_bart/bart.py

```diff
@@ -34,16 +38,24 @@ class BARTRV(RandomVariable):
     ndims_params = [2, 1, 0, 0, 1]
     dtype = "floatX"
     _print_name = ("BART", "\\operatorname{BART}")
+    all_trees = None
 
     def _supp_shape_from_params(self, dist_params, rep_param_idx=1, param_shapes=None):
-        return (self.X.shape[0],)
+        if isinstance(self.X, TensorSharedVariable):
+            shape = self.X.shape[0].eval()
```

Comment:
You should not have to `eval`, specially in this case as it is allowed to change.

Reply:
I assumed that too but if I don't eval i get "KeyError: data_X", maybe I am missing something else?

Reply:
Can you show a traceback?

Reply:
```python
KeyError                                  Traceback (most recent call last)
/home/osvaldo/proyectos/00_BM/dev/pymc/BART/preds/Bikes.ipynb Celda 7 in <cell line: 1>()
      [2](vscode-notebook-cell:/home/osvaldo/proyectos/00_BM/dev/pymc/BART/preds/Bikes.ipynb#W6sZmlsZQ%3D%3D?line=1) data_X = pm.MutableData("data_X", X)
      [3](vscode-notebook-cell:/home/osvaldo/proyectos/00_BM/dev/pymc/BART/preds/Bikes.ipynb#W6sZmlsZQ%3D%3D?line=2) μ = pmb.BART("μ", data_X, Y, m=2)
----> [4](vscode-notebook-cell:/home/osvaldo/proyectos/00_BM/dev/pymc/BART/preds/Bikes.ipynb#W6sZmlsZQ%3D%3D?line=3) y = pm.Poisson("y", np.abs(μ), observed=Y, shape=μ.shape)
      [5](vscode-notebook-cell:/home/osvaldo/proyectos/00_BM/dev/pymc/BART/preds/Bikes.ipynb#W6sZmlsZQ%3D%3D?line=4) idata_bikes = pm.sample(tune=100, draws=200, random_seed=RANDOM_SEED)

File ~/anaconda3/lib/python3.9/site-packages/pymc/distributions/distribution.py:438, in Discrete.__new__(cls, name, *args, **kwargs)
    435 if kwargs.get("transform", None):
    436     raise ValueError("Transformations for discrete distributions")
--> 438 return super().__new__(cls, name, *args, **kwargs)

File ~/anaconda3/lib/python3.9/site-packages/pymc/distributions/distribution.py:290, in Distribution.__new__(cls, name, rng, dims, initval, observed, total_size, transform, *args, **kwargs)
    287     elif observed is not None:
    288         kwargs["shape"] = tuple(observed.shape)
--> 290 rv_out = cls.dist(*args, **kwargs)
    292 rv_out = model.register_rv(
    293     rv_out,
    294     name,
   (...)
    299     initval=initval,
    300 )
    302 # add in pretty-printing support

File ~/anaconda3/lib/python3.9/site-packages/pymc/distributions/discrete.py:613, in Poisson.dist(cls, mu, *args, **kwargs)
    610 @classmethod
    611 def dist(cls, mu, *args, **kwargs):
    612     mu = at.as_tensor_variable(floatX(mu))
--> 613     return super().dist([mu], *args, **kwargs)

File ~/anaconda3/lib/python3.9/site-packages/pymc/distributions/distribution.py:369, in Distribution.dist(cls, dist_params, shape, **kwargs)
    367     ndim_supp = cls.rv_op(*dist_params, **kwargs).owner.op.ndim_supp
    368 create_size = find_size(shape=shape, size=size, ndim_supp=ndim_supp)
--> 369 rv_out = cls.rv_op(*dist_params, size=create_size, **kwargs)
    371 rv_out.logp = _make_nice_attr_error("rv.logp(x)", "pm.logp(rv, x)")
    372 rv_out.logcdf = _make_nice_attr_error("rv.logcdf(x)", "pm.logcdf(rv, x)")

File ~/anaconda3/lib/python3.9/site-packages/aesara/tensor/random/basic.py:930, in PoissonRV.__call__(self, lam, size, **kwargs)
    916 def __call__(self, lam=1.0, size=None, **kwargs):
    917     r"""Draw samples from a poisson distribution.
    918 
    919     Parameters
   (...)
    928 
    929     """
--> 930     return super().__call__(lam, size=size, **kwargs)

File ~/anaconda3/lib/python3.9/site-packages/aesara/tensor/random/op.py:279, in RandomVariable.__call__(self, size, name, rng, dtype, *args, **kwargs)
    278 def __call__(self, *args, size=None, name=None, rng=None, dtype=None, **kwargs):
--> 279     res = super().__call__(rng, size, dtype, *args, **kwargs)
    281     if name is not None:
    282         res.name = name

File ~/anaconda3/lib/python3.9/site-packages/aesara/graph/op.py:297, in Op.__call__(self, *inputs, **kwargs)
    255 r"""Construct an `Apply` node using :meth:`Op.make_node` and return its outputs.
    256 
    257 This method is just a wrapper around :meth:`Op.make_node`.
   (...)
    294 
    295 """
    296 return_list = kwargs.pop("return_list", False)
--> 297 node = self.make_node(*inputs, **kwargs)
    299 if config.compute_test_value != "off":
    300     compute_test_value(node)

File ~/anaconda3/lib/python3.9/site-packages/aesara/tensor/random/op.py:325, in RandomVariable.make_node(self, rng, size, dtype, *dist_params)
    320     raise TypeError(
    321         "The type of rng should be an instance of either RandomGeneratorType or RandomStateType"
    322     )
    324 shape = self._infer_shape(size, dist_params)
--> 325 _, bcast = infer_broadcastable(shape)
    326 dtype = self.dtype or dtype
    328 if dtype == "floatX":

File ~/anaconda3/lib/python3.9/site-packages/aesara/tensor/basic.py:1359, in infer_broadcastable(shape)
   1355     raise TypeError(f"Shapes must be scalar integers; got {s_as_str}")
   1357 sh = [check_type(as_tensor_variable(s, ndim=0)) for s in shape]
-> 1359 shape_fg = FunctionGraph(
   1360     outputs=sh,
   1361     features=[ShapeFeature()],
   1362     clone=True,
   1363 )
   1364 folded_shape = rewrite_graph(shape_fg, custom_rewrite=topo_constant_folding).outputs
   1366 bcast = tuple(getattr(s, "data", s) == 1 for s in folded_shape)

File ~/anaconda3/lib/python3.9/site-packages/aesara/graph/fg.py:153, in FunctionGraph.__init__(self, inputs, outputs, features, clone, update_mapping, **clone_kwds)
    150     self.add_input(in_var, check=False)
    152 for output in outputs:
--> 153     self.add_output(output, reason="init")
    155 self.profile = None
    156 self.update_mapping = update_mapping

File ~/anaconda3/lib/python3.9/site-packages/aesara/graph/fg.py:163, in FunctionGraph.add_output(self, var, reason, import_missing)
    161 """Add a new variable as an output to this `FunctionGraph`."""
    162 self.outputs.append(var)
--> 163 self.import_var(var, reason=reason, import_missing=import_missing)
    164 self.clients[var].append(("output", len(self.outputs) - 1))

File ~/anaconda3/lib/python3.9/site-packages/aesara/graph/fg.py:304, in FunctionGraph.import_var(self, var, reason, import_missing)
    302 # Imports the owners of the variables
    303 if var.owner and var.owner not in self.apply_nodes:
--> 304     self.import_node(var.owner, reason=reason, import_missing=import_missing)
    305 elif (
    306     var.owner is None
    307     and not isinstance(var, AtomicVariable)
    308     and var not in self.inputs
    309 ):
    310     from aesara.graph.null_type import NullType

File ~/anaconda3/lib/python3.9/site-packages/aesara/graph/fg.py:385, in FunctionGraph.import_node(self, apply_node, check, reason, import_missing)
    383         self.variables.add(input)
    384     self.add_client(input, (node, i))
--> 385 self.execute_callbacks("on_import", node, reason)

File ~/anaconda3/lib/python3.9/site-packages/aesara/graph/fg.py:725, in FunctionGraph.execute_callbacks(self, name, *args, **kwargs)
    723         continue
    724     tf0 = time.time()
--> 725     fn(self, *args, **kwargs)
    726     self.execute_callbacks_times[feature] += time.time() - tf0
    727 self.execute_callbacks_time += time.time() - t0

File ~/anaconda3/lib/python3.9/site-packages/aesara/tensor/rewriting/shape.py:587, in ShapeFeature.on_import(self, fgraph, node, reason)
    584         o_shapes[sh_idx] = tuple(new_shape)
    586 for r, s in zip(node.outputs, o_shapes):
--> 587     self.set_shape(r, s)

File ~/anaconda3/lib/python3.9/site-packages/aesara/tensor/rewriting/shape.py:365, in ShapeFeature.set_shape(self, r, s, override)
    363         shape_vars.append(constant(r.type.shape[i], dtype="int64"))
    364     else:
--> 365         shape_vars.append(self.unpack(s[i], r))
    366 assert all(
    367     not hasattr(r.type, "broadcastable")
    368     or not r.type.broadcastable[i]
   (...)
    371     for i in range(r.type.ndim)
    372 )
    373 self.shape_of[r] = tuple(shape_vars)

File ~/anaconda3/lib/python3.9/site-packages/aesara/tensor/rewriting/shape.py:320, in ShapeFeature.unpack(self, s_i, var)
    318         x = s_i.owner.inputs[0].owner.inputs[0]
    319         # x should already have been imported, and should be in shape_of.
--> 320         s_i = self.shape_of[x][i]
    322 if s_i.type.dtype in integer_dtypes:
    323     if getattr(s_i.type, "ndim", 0):

KeyError: data_X
```

Reply:
Looks like a `Aesara` bug, I'll try to investigate it .Is this PR very urgent?